core/src/main/java/fr/sncf/osrd/api/TimetableCacheManager.kt

```diff
@@ -0,0 +1,174 @@
+        return res
+    }
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    private fun fetchTrainRequirements(
+        infraId: String,
+        infra: RawInfra,
+        timetableId: TimetableId,
+    ): Flow<SpacingRequirement> = flow {
+        val firstPageTrainRequirements = getTrainPaginatedRequirements(infraId, timetableId, 1)
+        emitAll(
+            firstPageTrainRequirements.results
+                .flatMap {
+                    it.spacingRequirements.map { spacingReq ->
+                        SpacingRequirement.fromRJSWithAddedTime(
+                            spacingReq,
+                            infra,
+                            it.startTime.durationSinceEpoch(),
+                        )
+                    }
+                }
+                .asFlow()
+        )
+        emitAll(
+            (2..firstPageTrainRequirements.pageCount)
+                .asFlow()
+                // Limit the number of concurrent calls to the requirements endpoint.
+                .flatMapMerge(concurrency = 5) { page ->
+                    flow {
+                        val paginatedTrainRequirements =
+                            getTrainPaginatedRequirements(infraId, timetableId, page)
+                        paginatedTrainRequirements.results
+                            .flatMap {
+                                it.spacingRequirements.map { spacingReq ->
+                                    SpacingRequirement.fromRJSWithAddedTime(
+                                        spacingReq,
+                                        infra,
+                                        it.startTime.durationSinceEpoch(),
+                                    )
+                                }
+                            }
+                            .forEach { emit(it) }
+                    }
+                }
+        )
+    }
+
+    private fun getTrainPaginatedRequirements(
+        infraId: String,
+        timetableId: TimetableId,
+        page: Int,
+    ): PaginatedRequirements {
+        val endpointPath = "timetable/$timetableId/requirements/"
+        val request =
+            buildRequest(endpointPath, "infra_id=$infraId&page=$page&page_size=$PAGE_SIZE")
+        val response = httpClient.newCall(request).execute()
+        if (!response.isSuccessful) throw UnexpectedHttpResponse(response)
+        return paginatedRequirementsAdapter.fromJson(response.body.source())!!
+    }
+
+    private data class PaginatedRequirements(
+        @Json(name = "page_count") val pageCount: Int,
+        val results: List<TrainRequirementsById>,
+    )
+
+    private val paginatedRequirementsAdapter: JsonAdapter<PaginatedRequirements> =
+        Moshi.Builder()
+            .addLast(UnitAdapterFactory())
+            .addLast(KotlinJsonAdapterFactory())
+            .build()
+            .adapter(PaginatedRequirements::class.java)
+}
+
+val EPOCH_ZONED: ZonedDateTime = Instant.EPOCH.atZone(java.time.ZoneId.of("UTC"))
```

Comment:
Seconds since epoch is a large value, and times are stored as doubles. We need to make sure we keep enough precision. I'll run tests.